libs/api/ebpf_api.cpp

```diff
@@ -673,3 +676,401 @@ ebpf_api_map_info_free(_In_ const uint16_t map_count, _In_count_(map_count) cons
     ebpf_map_information_array_free(map_count, const_cast<ebpf_map_information_t*>(map_info));
 }
 
+void
+clean_up_ebpf_program(ebpf_program_t* program)
+{
+    if (program == nullptr) {
+        return;
+    }
+    if (program->fd != 0) {
+        _ebpf_programs.erase(program->fd);
+    }
+    if (program->handle != ebpf_handle_invalid) {
+        CloseHandle(program->handle);
+    }
+    free(program->byte_code);
+    free(program->program_name);
+    free(program->section_name);
+
+    free(program);
+}
+
+void
+clean_up_ebpf_programs(ebpf_list_entry_t& programs)
+{
+    ebpf_list_entry_t* entry;
+    ebpf_program_t* program;
+    while ((entry = ebpf_list_remove_head_entry(&programs)) != &programs) {
+        program = CONTAINING_RECORD(entry, ebpf_program_t, list_entry);
+        clean_up_ebpf_program(program);
+    }
+}
+
+void
+clean_up_ebpf_map(ebpf_map_t* map)
+{
+    if (map->map_fd != 0) {
+        _ebpf_maps.erase(map->map_fd);
+    }
+    if (map->map_handle != ebpf_handle_invalid) {
+        CloseHandle(map->map_handle);
+    }
+
+    free(map);
+}
+
+void
+clean_up_ebpf_maps(ebpf_list_entry_t& maps)
+{
+    ebpf_list_entry_t* entry;
+    ebpf_map_t* map;
+    while ((entry = ebpf_list_remove_head_entry(&maps)) != &maps) {
+        map = CONTAINING_RECORD(entry, ebpf_map_t, list_entry);
+        clean_up_ebpf_map(map);
+    }
+}
+
+static void
+_clean_up_ebpf_object(ebpf_object_t* object)
+{
+    if (object != nullptr) {
+        clean_up_ebpf_programs(object->programs);
+        clean_up_ebpf_maps(object->maps);
+
+        free(object);
+    }
+}
+
+static void
+_remove_ebpf_object_from_globals(ebpf_object_t* object)
+{
+    for (int i = 0; i < _ebpf_objects.size(); i++) {
+        if (_ebpf_objects[i] == object) {
+            // _ebpf_objects.erase(_ebpf_objects.begin() + i);
+            _ebpf_objects[i] = nullptr;
+            break;
+        }
+    }
+}
+
+static void
+_clean_up_ebpf_objects()
+{
+    for (auto& object : _ebpf_objects) {
+        _clean_up_ebpf_object(object);
+    }
+
+    _ebpf_objects.resize(0);
+
+    assert(_ebpf_programs.size() == 0);
+    assert(_ebpf_maps.size() == 0);
+}
+
+static void
+_initialize_map(ebpf_map_t* map, ebpf_object_t* object, const map_cache_t& map_cache)
+{
+    map->object = object;
+    map->map_handle = (ebpf_handle_t)map_cache.handle;
+    map->map_fd = map_cache.ebpf_map_descriptor.original_fd;
+    map->map_defintion.type = (ebpf_map_type_t)map_cache.ebpf_map_descriptor.type;
+    map->map_defintion.key_size = map_cache.ebpf_map_descriptor.key_size;
+    map->map_defintion.value_size = map_cache.ebpf_map_descriptor.value_size;
+    map->map_defintion.max_entries = map_cache.ebpf_map_descriptor.max_entries;
+    map->pinned = false;
+    map->pin_path = nullptr;
+}
+
+static ebpf_result_t
+_initialize_ebpf_object_from_elf(
```

Comment:
I think SAL annotation would help throughout this file

Reply:
Added SAL annotations.